### ISSUE TITLE
DocAPI: Update records set with an option to replace the whole table

### ISSUE THREAD

**Issue by fflorent** [gristlabs/grist-core#626]
As a developer, I would like to replace the whole records of a table with a new set of records through the API so I can remove records that are not in my records set (related to issue #416)

The [PUT verb on columns (aka `replaceColumns`)](https://github.com/gristlabs/grist-help/blob/ded498016cf5ba2c145d825cc42ed2a73f8f69bf/api/grist.yml#L905) implements a `replaceall` parameter for that purpose. I suggest to do the same for records.

**Comment by fflorent** [comment]
Regarding the technical part, I would suggest to add an option `replace_all` to the [`BulkAddOrUpdateRecord` python function](https://github.com/gristlabs/grist-core/blob/9df62e3d81163df08622b16fa82806ab8050ae17/sandbox/grist/useractions.py#L924).

I started to implement that change here: https://github.com/incubateur-territoires/grist-core/commit/89727dcecd6f41c8251b7287e38013b382fc6546

However, I see that `ops.upsert` calls BulkAddOrUpdateRecord several times with this reason explained in comments:
https://github.com/gristlabs/grist-core/blob/9df62e3d81163df08622b16fa82806ab8050ae17/app/plugin/TableOperationsImpl.ts#L64-L67

I don't have the reasons of why filling array with null values would be bad though.
So I don't know what is the right way to handle this. Do you have some advice/hint to give?

**Comment by fflorent** [comment]
Maybe @alexmojaki you could help me on this one? :pray:

**Comment by alexmojaki** [comment]
> I don't have the reasons of why filling array with null values would be bad though.

Because that would set random cell values to null which the user probably doesn't want.

`ops.upsert` with multiple records having different keys in require/fields seems like a strange thing to do, the code is just being cautious. How do you imagine using `PUT /records`? Wouldn't all the records have the same shape?

Anyway, have you considered `POST /apply` with the `ReplaceTableData` user action?

**Comment by fflorent** [comment]
> Anyway, have you considered POST /apply with the ReplaceTableData user action?

ReplaceTableData seems to do the job. Also I was not aware that this endpoint (`/apply`) existed, as it seems that it is not documented in the API documentation (is it intended or would you welcome a PR for that in grist-help?)

**Comment by alexmojaki** [comment]
User actions themselves aren't documented, so I don't know if it would be helpful. @paulfitz WDYT?

**Comment by paulfitz** [comment]
The undocumented `/apply` endpoint could be a *lot* of effort to document effectively. It is there so that any changes that can be done via the web client are also possible (but not necessarily easy) via the rest api. When we've had time for API improvements, we've been spending it on adding cleaner, more self-contained endpoints like `/records` and `/columns` to do the most common operations, rather than on documenting `/apply`. The endpoint is stable though, and won't be going away.

TLDR a PR that documents `/apply` well would be accepted :-)
